### PR TITLE
Add Call block to program MAST

### PR DESCRIPTION
**Pull request by bobbinth**
This PR adds a `Call` block to program MAST. This includes:

* Added a code block table to `Program` struct. This creates a sort of a "virtual table" structure within a program such that internal nodes in the program MAST can refer to code blocks in the table. This is needed for handling call and syscall blocks, but could come in handing in other places too.
* Updated assembly parsing to handle `call` statements. These look similar to `exec` statements, but semantically are different. While `exec` statements inline procedures into the main program MAST, `call` statements add an entry to the program's code block table and and adde a `Call` block node to the main program MAST.
* Updated processor to handle call blocks. This includes updating the decoder to enable enable basic processing of call blocks.
* Added a tiny test for to check that call blocks are executed correctly.

Surprisingly, these changes also speed up program compilation and execution time by 4% - 6%.

This PR does not yet include:
* Memory context isolation.
* Stack depth manipulation.
* Documentation updates.

These will be addressed in future PRs.

### CODE REVIEW

**Review by grjte — CHANGES_REQUESTED** [review]
I may not have fully understood what's supposed to be happening, so I left a comment inline for clarification.

Also, while trying to get a better understanding of this I added a `prove_and_verify` test to the end of the simple `local_fn_call` test you created, and it failed.

Is this expected to fail at this point?

```
#[test]
fn local_fn_call() {
    let source = "
        proc.foo
            add
        end

        begin
            push.1
            push.2
            call.foo
        end";

    let test = build_test!(source);
    test.expect_stack(&[3]);

    test.prove_and_verify(vec![], 1, false);
}
```


Note: I tried testing the other control flow ops in the same way and `flow_control::counter_controlled_loop` also fails. The rest are fine. This is true before these changes as well, as expected. (In the integration tests flow control module)

**Comment by bobbinth** [comment]
> Is this expected to fail at this point?

So, this fails because the stack at the end is not 16. I've changed it to the below and it seems to work fine:
```Rust
#[test]
fn local_fn_call() {
    let source = "
        proc.foo
            add
        end

        begin
            exec.foo
        end";

    let test = build_test!(source, &[1, 2]);
    test.expect_stack(&[3]);

    test.prove_and_verify(vec![1, 2], 1, false);
}
```

File: processor/src/decoder/mod.rs [hunk 1]
```diff
@@ -407,6 +438,23 @@ impl Decoder {
...
+        let parent_addr = self.block_stack.push(addr, BlockType::Call);
+        self.trace
+            .append_block_start(parent_addr, Operation::Join, fn_hash, [ZERO; 4]);
+
+        // mark this cycle as the cycle at which a new CALL block began execution (this affects
+        // block stack and block hash tables). A CALL block has only a single child.
+        self.aux_hints
+            .block_started(clk, self.block_stack.peek(), Some(fn_hash), None);
+
+        self.debug_info.append_operation(Operation::Join);
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply processor/src/decoder/mod.rs:455]: That's a copy-paste error on my end. Fixed!

**Comment by grjte** [comment]
> So, this fails because the stack at the end is not 16. I've changed it to the below and it seems to work fine:

Ah great, good point, should've realized that 😅 I'll be happy to see the end of that requirement!

File: core/src/program/mod.rs [hunk 2]
```diff
@@ -43,10 +52,47 @@ impl Program {
...
         write!(f, "begin {} end", self.root)
     }
 }
+
+// CODE BLOCK TABLE
+// ================================================================================================
+
+/// A map of code block hashes to their underlying code blocks.
+///
+/// This table is used to hold code blocks which a referenced from the program MAST but are
```
Comment (grjte) [inline core/src/program/mod.rs:73]: nit: this still has a typo (there were 2 before). It says "a referenced" instead of "are referenced"

**Review by grjte — APPROVED** [review]
Looks good now, thank you! I just left one nit inline for a typo in a doc comment

File: core/src/program/mod.rs (hunk 2 above)
Comment (bobbinth) [reply core/src/program/mod.rs]: Ah yes - sorry! fixed!

**Pull request merged**